crud/common/sharding_key.lua

```diff
@@ -0,0 +1,122 @@
+    dev_checks('table', 'string', 'table')
+
+    local fieldnos = {}
+    local fieldno_map = utils.get_format_fieldno_map(space_format)
+    for _, field_name in ipairs(sharding_key_def) do
+        local fieldno = fieldno_map[field_name]
+        if fieldno == nil then
+            return nil, GetShardingKeyFieldnosError:new(
+                "No such field (%s) in a space format (%s)", field_name, space_name)
+        end
+        table.insert(fieldnos, {fieldno = fieldno})
+    end
+
+    return {parts = fieldnos}
+end
+
+-- It's a public method to access it in tests.
+function sharding_key_module.fetch_on_storage()
+    local sharding_key_def = {}
+    local sharding_key_space = box.space._ddl_sharding_key
+    if sharding_key_space == nil then
+        return sharding_key_def
+    end
+
+    local SPACE_NAME_FIELDNO = 1
+    local SPACE_SHARDING_KEY_FIELDNO = 2
+    for _, tuple in ipairs(sharding_key_space:select()) do
+        local space_name = tuple[SPACE_NAME_FIELDNO]
+        local key = tuple[SPACE_SHARDING_KEY_FIELDNO]
+        sharding_key_def[space_name] = key
+    end
+
+    return sharding_key_def
+end
+
+local sharding_key_fetch_in_progress = false
+local sharding_key_fetch_cond = fiber.cond()
+
+-- Under high load we may get a case when many fibers will fetch sharding keys
+-- from storages. It is not good from performance point of view.
+-- concurrent_fetch() wraps a fetch_on_storage() to limit a number of fibers
+-- that fetches a sharding key by a single one, other fibers will wait for a
+-- FETCH_SHARDING_KEY_TIMEOUT and then make one more attempt to fetch sharding
+-- key if a cache is still empty.
+local function concurrent_fetch(replicaset)
+    dev_checks('table')
+
+    if sharding_key_fetch_in_progress == true and
+       not sharding_key_fetch_cond:wait(const.FETCH_SHARDING_KEY_TIMEOUT) then
+        return nil, FetchShardingKeyError:new(
+                        'Waiting for sharding key to be fetched is timed out')
+    end
+
+    local sharding_key_def = {}
+    if sharding_key_fetch_in_progress == false then
+        sharding_key_fetch_in_progress = true
+        local err
+        sharding_key_def, err = replicaset:call(FETCH_FUNC_NAME, {},
+                                                {timeout = const.FETCH_SHARDING_KEY_TIMEOUT})
+        if err ~= nil then
+            return nil, err
+        end
+        sharding_key_fetch_cond:broadcast()
+        sharding_key_fetch_in_progress = false
+    end
+
+    return sharding_key_def
+end
+
+-- Get sharding key definitions (field names) for all spaces or a certain space
+-- (when space_name is not nil). Function uses a cache that organized as a table in
+-- following format:
+-- {
+--     'space_name_1' = {'name', 'age'},
+--     'space_name_2' = {'age'},
+--     ...
+-- }
+-- Return an sharding key or nil when space_name is specified or a whole map
+-- with all sharding keys.
+function sharding_key_module.fetch_on_router(space_name)
+    dev_checks('string')
```

Comment:
Is it mandatory or optional?

Reply:
I mean, the API description marks it as optional, the code checks whether it is `nil`, but here we require it to be non-nil.

Reply:
Mandatory. Updated comment.

```
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -96,8 +96,7 @@ end
 --     'space_name_2' = {'age'},
 --     ...
 -- }
--- Return an sharding key or nil when space_name is specified or a whole map
--- with all sharding keys.
+-- Return a sharding key definition (table with field names) or nil.
 function sharding_key_module.fetch_on_router(space_name)
     dev_checks('string')
 
```

Reply:
However we still check it against `nil`:

```lua
if space_name ~= nil then
    return cache.sharding_key_def[space_name]
end
```

Reply:
Fixed.

```lua
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -109,11 +109,8 @@ function sharding_key_module.fetch_on_router(space_name)
             return nil, err
         end
     end
-    if space_name ~= nil then
-        return cache.sharding_key_def[space_name]
-    end
 
-    return cache.sharding_key_def
+    return cache.sharding_key_def[space_name]
 end
 
 function sharding_key_module.update_sharding_keys_cache(space_name)
```